### PR TITLE
HOTT-1918: Support groups in excluded geographical areas

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1918

### What?

I have added/removed/altered:

- [x] Add support for handling groups in measure geographical area exclusions
- [x] Adds spec coverage for new behaviour

### Why?

I am doing this because:

- A customer noticed that this wasn't supported in the tariff
- On a side note we're not actually setting the group correctly for the EU in the measure excluded geographical areas (this will need to be fixed separately)

### CODE REVIEW

File: app/models/measure.rb [hunk 1]
```diff
@@ -413,8 +409,15 @@ def order_number
     end
   end
 
-  def relevant_for_country?(country_id)
-    return false if measure_excluded_geographical_areas.map(&:excluded_geographical_area).include?(country_id)
+  def relevant_for_country?(country)
+    country_and_memberships = country.included_geographical_areas.pluck(:geographical_area_id) << country.geographical_area_id
```
Comment (willfish) [inline app/models/measure.rb:413]: Push this into a method on geographical area

File: app/models/geographical_area.rb [hunk 2]
```diff
@@ -80,4 +89,10 @@ def areas
   def id
     geographical_area_id
   end
+
+  private
+
+  def referenced_id
```
Comment (willfish) [inline app/models/geographical_area.rb:95]: We could do a migration to enable a reference but I'd rather hard code this in a hash for now as it very much could be yagni

File: app/models/rules_of_origin/scheme.rb [hunk 3]
```diff
@@ -43,7 +43,7 @@ def proofs
     end
 
     def articles
-      @articles ||= Article.for_scheme(self)
+      @articles ||= RulesOfOrigin::Article.for_scheme(self)
```
Comment (willfish) [inline app/models/rules_of_origin/scheme.rb:46]: This fixes intermittent zeitwerk loading issues

File: spec/factories/geographical_area_factory.rb [hunk 4]
```diff
@@ -8,14 +8,6 @@
     validity_start_date   { 3.years.ago.beginning_of_day }
     validity_end_date     { nil }
 
-    trait :fifteen_years do
```
Comment (willfish) [inline spec/factories/geographical_area_factory.rb:11]: These are unused

File: app/services/meursing_measure_finder_service.rb [hunk 5]
```diff
@@ -28,6 +28,6 @@ def call
   attr_reader :root_measure, :additional_code_id
 
   def relevant_for_country?(measure)
-    measure.relevant_for_country?(root_measure.geographical_area_id)
+    measure.relevant_for_country?(root_measure.geographical_area)
```
Comment (jebw) [inline app/services/meursing_measure_finder_service.rb:31]: This shifts from using an attribute on `root_measure` to a record lookup - is this called in a loop anyway, ie will it introduce an N+1?

File: app/models/geographical_area.rb [hunk 6]
```diff
@@ -48,6 +49,14 @@ def geographical_area_description
     ds.with_actual(GeographicalAreaMembership).order(Sequel.asc(:geographical_area_id))
   end
 
+  def candidate_excluded_geographical_area_ids
+    @candidate_excluded_geographical_area_ids ||= contained_geographical_areas.pluck(:geographical_area_id) << geographical_area_id
```
[the remark this answers is not in the document]
Comment (willfish) [reply app/models/geographical_area.rb]: Ah good spot

File: app/services/meursing_measure_finder_service.rb (hunk 5 above)
Comment (willfish) [reply app/services/meursing_measure_finder_service.rb]: Resolved this now :)

**Pull request merged**